{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c2e34671-b50d-447a-8951-4d4f71a14554",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5db8c2e8-1fce-4ed6-b0b5-8e08bde45c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "candidate_seq_f = '../data3/interim/candidate_seqs.faa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a36ce5fa-a330-40e6-a00e-1c117c859a48",
   "metadata": {},
   "outputs": [],
   "source": [
    "search_out = '../data3/interim/candiate_seq_mmseqs_self_search.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a158c2bc-3582-4398-a539-2848a0795184",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../data3/interim/candiate_seq_mmseqs_self_search.txt exists and will be overwritten\n",
      "easy-search ../data3/interim/candidate_seqs.faa ../data3/interim/candidate_seqs.faa ../data3/interim/candiate_seq_mmseqs_self_search.txt /state/partition1/slurm_tmp/1049207.0.0 --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov --threads 48 -s 7.5 \n",
      "\n",
      "MMseqs Version:                        \t13.45111\n",
      "Substitution matrix                    \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Add backtrace                          \tfalse\n",
      "Alignment mode                         \t3\n",
      "Alignment mode                         \t0\n",
      "Allow wrapped scoring                  \tfalse\n",
      "E-value threshold                      \t0.001\n",
      "Seq. id. threshold                     \t0\n",
      "Min alignment length                   \t0\n",
      "Seq. id. mode                          \t0\n",
      "Alternative alignments                 \t0\n",
      "Coverage threshold                     \t0\n",
      "Coverage mode                          \t0\n",
      "Max sequence length                    \t65535\n",
      "Compositional bias                     \t1\n",
      "Max reject                             \t2147483647\n",
      "Max accept                             \t2147483647\n",
      "Include identical seq. id.             \tfalse\n",
      "Preload mode                           \t0\n",
      "Pseudo count a                         \t1\n",
      "Pseudo count b                         \t1.5\n",
      "Score bias                             \t0\n",
      "Realign hits                           \tfalse\n",
      "Realign score bias                     \t-0.2\n",
      "Realign max seqs                       \t2147483647\n",
      "Gap open cost                          \tnucl:5,aa:11\n",
      "Gap extension cost                     \tnucl:2,aa:1\n",
      "Zdrop                                  \t40\n",
      "Threads                                \t48\n",
      "Compressed                             \t0\n",
      "Verbosity                              \t3\n",
      "Seed substitution matrix               \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                            \t7.5\n",
      "k-mer length                           \t0\n",
      "k-score                                \t2147483647\n",
      "Alphabet size                          \tnucl:5,aa:21\n",
      "Max results per query                  \t300\n",
      "Split database                         \t0\n",
      "Split mode                             \t2\n",
      "Split memory limit                     \t0\n",
      "Diagonal scoring                       \ttrue\n",
      "Exact k-mer matching                   \t0\n",
      "Mask residues                          \t1\n",
      "Mask lower case residues               \t0\n",
      "Minimum diagonal score                 \t15\n",
      "Spaced k-mers                          \t1\n",
      "Spaced k-mer pattern                   \t\n",
      "Local temporary path                   \t\n",
      "Rescore mode                           \t0\n",
      "Remove hits by seq. id. and coverage   \tfalse\n",
      "Sort results                           \t0\n",
      "Mask profile                           \t1\n",
      "Profile E-value threshold              \t0.001\n",
      "Global sequence weighting              \tfalse\n",
      "Allow deletions                        \tfalse\n",
      "Filter MSA                             \t1\n",
      "Maximum seq. id. threshold             \t0.9\n",
      "Minimum seq. id.                       \t0\n",
      "Minimum score per column               \t-20\n",
      "Minimum coverage                       \t0\n",
      "Select N most diverse seqs             \t1000\n",
      "Min codons in orf                      \t30\n",
      "Max codons in length                   \t32734\n",
      "Max orf gaps                           \t2147483647\n",
      "Contig start mode                      \t2\n",
      "Contig end mode                        \t2\n",
      "Orf start mode                         \t1\n",
      "Forward frames                         \t1,2,3\n",
      "Reverse frames                         \t1,2,3\n",
      "Translation table                      \t1\n",
      "Translate orf                          \t0\n",
      "Use all table starts                   \tfalse\n",
      "Offset of numeric ids                  \t0\n",
      "Create lookup                          \t0\n",
      "Add orf stop                           \tfalse\n",
      "Overlap between sequences              \t0\n",
      "Sequence split mode                    \t1\n",
      "Header split mode                      \t0\n",
      "Chain overlapping alignments           \t0\n",
      "Merge query                            \t1\n",
      "Search type                            \t0\n",
      "Search iterations                      \t1\n",
      "Start sensitivity                      \t4\n",
      "Search steps                           \t1\n",
      "Exhaustive search mode                 \tfalse\n",
      "Filter results during exhaustive search\t0\n",
      "Strand selection                       \t1\n",
      "LCA search mode                        \tfalse\n",
      "Disk space limit                       \t0\n",
      "MPI runner                             \t\n",
      "Force restart with latest tmp          \tfalse\n",
      "Remove temporary files                 \ttrue\n",
      "Alignment format                       \t0\n",
      "Format alignment output                \tquery,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov\n",
      "Database output                        \tfalse\n",
      "Overlap threshold                      \t0\n",
      "Database type                          \t0\n",
      "Shuffle input database                 \ttrue\n",
      "Createdb mode                          \t0\n",
      "Write lookup file                      \t0\n",
      "Greedy best hits                       \tfalse\n",
      "\n",
      "createdb ../data3/interim/candidate_seqs.faa /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/query --dbtype 0 --shuffle 1 --createdb-mode 0 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[====================\n",
      "Time for merging to query_h: 0h 0m 0s 39ms\n",
      "Time for merging to query: 0h 0m 0s 82ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 745ms\n",
      "createdb ../data3/interim/candidate_seqs.faa /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/target --dbtype 0 --shuffle 1 --createdb-mode 0 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[====================\n",
      "Time for merging to target_h: 0h 0m 0s 34ms\n",
      "Time for merging to target: 0h 0m 0s 81ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 628ms\n",
      "Create directory /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/search_tmp\n",
      "search /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/query /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/target /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/result /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/search_tmp --alignment-mode 3 --threads 48 -s 7.5 --remove-tmp-files 1 \n",
      "\n",
      "prefilter /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/query /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/target /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/search_tmp/12147251001121954483/pref_0 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 300 --split 0 --split-mode 2 --split-memory-limit 0 -c 0 --cov-mode 0 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 48 --compressed 0 -v 3 -s 7.5 \n",
      "\n",
      "Query database size: 200655 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 200655 type: Aminoacid\n",
      "Index table k-mer threshold: 96 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 200.65K 0s 655ms\n",
      "Index table: Masked residues: 1188108\n",
      "Index table: fill\n",
      "[=================================================================] 200.65K 0s 449ms\n",
      "Index statistics\n",
      "Entries:          69957238\n",
      "DB size:          888 MB\n",
      "Avg k-mer size:   1.093082\n",
      "Top 10 k-mers\n",
      "    DLLDMP\t1611\n",
      "    LAMHRT\t1453\n",
      "    GEGGVV\t1182\n",
      "    TSAPAG\t1075\n",
      "    FLNSIR\t953\n",
      "    GFVAVR\t911\n",
      "    EARGGR\t851\n",
      "    GGGGIS\t797\n",
      "    RGAVEV\t678\n",
      "    DLIDMP\t633\n",
      "Time for index table init: 0h 0m 1s 569ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 96\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 200655\n",
      "Target db start 1 to 200655\n",
      "[=================================================================] 200.65K 2m 42s 524ms\n",
      "\n",
      "1344.733185 k-mers per position\n",
      "625013 DB matches per sequence\n",
      "6950 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "299 sequences passed prefiltering per query sequence\n",
      "300 median result list length\n",
      "3 sequences with 0 size result lists\n",
      "Time for merging to pref_0: 0h 0m 0s 55ms\n",
      "Time for processing: 0h 2m 45s 878ms\n",
      "align /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/query /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/target /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/search_tmp/12147251001121954483/pref_0 /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/result --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 48 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 200655 type: Aminoacid\n",
      "Target database size: 200655 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 200.65K 4m 32s 212ms\n",
      "Time for merging to result: 0h 0m 0s 63ms\n",
      "60025388 alignments calculated\n",
      "19047244 sequence pairs passed the thresholds (0.317320 of overall calculated)\n",
      "94.925339 hits per query sequence\n",
      "Time for processing: 0h 4m 33s 25ms\n",
      "rmdb /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/search_tmp/12147251001121954483/pref_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 83ms\n",
      "rmdb /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/search_tmp/12147251001121954483/aln_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/search_tmp/12147251001121954483/input_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/search_tmp/12147251001121954483/aln_merge -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "../data3/interim/candiate_seq_mmseqs_self_search.txt exists and will be overwritten\n",
      "convertalis /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/query /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/target /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/result ../data3/interim/candiate_seq_mmseqs_self_search.txt --sub-mat nucl:nucleotide.out,aa:blosum62.out --format-mode 0 --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov --translation-table 1 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --db-output 0 --db-load-mode 0 --search-type 0 --threads 48 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 200.65K 1s 783ms\n",
      "Time for merging to candiate_seq_mmseqs_self_search.txt: 0h 0m 3s 251ms\n",
      "Time for processing: 0h 0m 8s 411ms\n",
      "rmdb /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 106ms\n",
      "rmdb /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/target -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "rmdb /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/target_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/query -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "rmdb /state/partition1/slurm_tmp/1049207.0.0/18145387415082206335/query_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "\n",
      "CPU times: user 58.7 ms, sys: 4.76 ms, total: 63.4 ms\n",
      "Wall time: 7min 30s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "os.system(' '.join(['conda run -n beaker', \n",
    "                    'mmseqs easy-search', \n",
    "                    candidate_seq_f, \n",
    "                    candidate_seq_f, \n",
    "                    search_out,\n",
    "                    '$TMPDIR',\n",
    "                    '--format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov', \n",
    "                    '--threads 48',\n",
    "                    '-s 7.5']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56972ff1-6c8a-4e9b-a5ca-3d3a9e399ae8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "beaker",
   "language": "python",
   "name": "beaker"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
