{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9e1c6f08-ea53-4300-9e5e-f8a253558002",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from core import read_mmseqs_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "83c01686-1aa6-4a55-a006-dad5508655ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_seq_info = pd.read_parquet('../data3/interim/model_seq_info.pq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5dd55e8-a5a3-4190-ad27-1bfcd7d17fee",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_seq_f = '../data3/interim/model_seqs.faa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b4a15a18-71ee-47a2-9be9-4aec4657b546",
   "metadata": {},
   "outputs": [],
   "source": [
    "mmseqs_search_out = '../data3/interim/model_seq_search.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "18d7e214-d104-459e-ac4e-292a9ba9a512",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../data3/interim/model_seq_search.txt exists and will be overwritten\n",
      "easy-search ../data3/interim/model_seqs.faa ../data3/interim/model_seqs.faa ../data3/interim/model_seq_search.txt ../data/tmp --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov --threads 48 -s 7.5 \n",
      "\n",
      "MMseqs Version:                        \t13.45111\n",
      "Substitution matrix                    \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Add backtrace                          \tfalse\n",
      "Alignment mode                         \t3\n",
      "Alignment mode                         \t0\n",
      "Allow wrapped scoring                  \tfalse\n",
      "E-value threshold                      \t0.001\n",
      "Seq. id. threshold                     \t0\n",
      "Min alignment length                   \t0\n",
      "Seq. id. mode                          \t0\n",
      "Alternative alignments                 \t0\n",
      "Coverage threshold                     \t0\n",
      "Coverage mode                          \t0\n",
      "Max sequence length                    \t65535\n",
      "Compositional bias                     \t1\n",
      "Max reject                             \t2147483647\n",
      "Max accept                             \t2147483647\n",
      "Include identical seq. id.             \tfalse\n",
      "Preload mode                           \t0\n",
      "Pseudo count a                         \t1\n",
      "Pseudo count b                         \t1.5\n",
      "Score bias                             \t0\n",
      "Realign hits                           \tfalse\n",
      "Realign score bias                     \t-0.2\n",
      "Realign max seqs                       \t2147483647\n",
      "Gap open cost                          \tnucl:5,aa:11\n",
      "Gap extension cost                     \tnucl:2,aa:1\n",
      "Zdrop                                  \t40\n",
      "Threads                                \t48\n",
      "Compressed                             \t0\n",
      "Verbosity                              \t3\n",
      "Seed substitution matrix               \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                            \t7.5\n",
      "k-mer length                           \t0\n",
      "k-score                                \t2147483647\n",
      "Alphabet size                          \tnucl:5,aa:21\n",
      "Max results per query                  \t300\n",
      "Split database                         \t0\n",
      "Split mode                             \t2\n",
      "Split memory limit                     \t0\n",
      "Diagonal scoring                       \ttrue\n",
      "Exact k-mer matching                   \t0\n",
      "Mask residues                          \t1\n",
      "Mask lower case residues               \t0\n",
      "Minimum diagonal score                 \t15\n",
      "Spaced k-mers                          \t1\n",
      "Spaced k-mer pattern                   \t\n",
      "Local temporary path                   \t\n",
      "Rescore mode                           \t0\n",
      "Remove hits by seq. id. and coverage   \tfalse\n",
      "Sort results                           \t0\n",
      "Mask profile                           \t1\n",
      "Profile E-value threshold              \t0.001\n",
      "Global sequence weighting              \tfalse\n",
      "Allow deletions                        \tfalse\n",
      "Filter MSA                             \t1\n",
      "Maximum seq. id. threshold             \t0.9\n",
      "Minimum seq. id.                       \t0\n",
      "Minimum score per column               \t-20\n",
      "Minimum coverage                       \t0\n",
      "Select N most diverse seqs             \t1000\n",
      "Min codons in orf                      \t30\n",
      "Max codons in length                   \t32734\n",
      "Max orf gaps                           \t2147483647\n",
      "Contig start mode                      \t2\n",
      "Contig end mode                        \t2\n",
      "Orf start mode                         \t1\n",
      "Forward frames                         \t1,2,3\n",
      "Reverse frames                         \t1,2,3\n",
      "Translation table                      \t1\n",
      "Translate orf                          \t0\n",
      "Use all table starts                   \tfalse\n",
      "Offset of numeric ids                  \t0\n",
      "Create lookup                          \t0\n",
      "Add orf stop                           \tfalse\n",
      "Overlap between sequences              \t0\n",
      "Sequence split mode                    \t1\n",
      "Header split mode                      \t0\n",
      "Chain overlapping alignments           \t0\n",
      "Merge query                            \t1\n",
      "Search type                            \t0\n",
      "Search iterations                      \t1\n",
      "Start sensitivity                      \t4\n",
      "Search steps                           \t1\n",
      "Exhaustive search mode                 \tfalse\n",
      "Filter results during exhaustive search\t0\n",
      "Strand selection                       \t1\n",
      "LCA search mode                        \tfalse\n",
      "Disk space limit                       \t0\n",
      "MPI runner                             \t\n",
      "Force restart with latest tmp          \tfalse\n",
      "Remove temporary files                 \ttrue\n",
      "Alignment format                       \t0\n",
      "Format alignment output                \tquery,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov\n",
      "Database output                        \tfalse\n",
      "Overlap threshold                      \t0\n",
      "Database type                          \t0\n",
      "Shuffle input database                 \ttrue\n",
      "Createdb mode                          \t0\n",
      "Write lookup file                      \t0\n",
      "Greedy best hits                       \tfalse\n",
      "\n",
      "createdb ../data3/interim/model_seqs.faa ../data/tmp/18076087456686426716/query --dbtype 0 --shuffle 1 --createdb-mode 0 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[===================\n",
      "Time for merging to query_h: 0h 0m 0s 419ms\n",
      "Time for merging to query: 0h 0m 0s 266ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 2s 231ms\n",
      "createdb ../data3/interim/model_seqs.faa ../data/tmp/18076087456686426716/target --dbtype 0 --shuffle 1 --createdb-mode 0 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[===================\n",
      "Time for merging to target_h: 0h 0m 0s 357ms\n",
      "Time for merging to target: 0h 0m 0s 266ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 2s 277ms\n",
      "Create directory ../data/tmp/18076087456686426716/search_tmp\n",
      "search ../data/tmp/18076087456686426716/query ../data/tmp/18076087456686426716/target ../data/tmp/18076087456686426716/result ../data/tmp/18076087456686426716/search_tmp --alignment-mode 3 --threads 48 -s 7.5 --remove-tmp-files 1 \n",
      "\n",
      "prefilter ../data/tmp/18076087456686426716/query ../data/tmp/18076087456686426716/target ../data/tmp/18076087456686426716/search_tmp/2581302743047066002/pref_0 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 300 --split 0 --split-mode 2 --split-memory-limit 0 -c 0 --cov-mode 0 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 48 --compressed 0 -v 3 -s 7.5 \n",
      "\n",
      "Query database size: 190804 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 190804 type: Aminoacid\n",
      "Index table k-mer threshold: 96 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 190.80K 0s 605ms\n",
      "Index table: Masked residues: 1040848\n",
      "Index table: fill\n",
      "[=================================================================] 190.80K 0s 463ms\n",
      "Index statistics\n",
      "Entries:          65817964\n",
      "DB size:          864 MB\n",
      "Avg k-mer size:   1.028406\n",
      "Top 10 k-mers\n",
      "    DLLDMP\t1395\n",
      "    LAMHRT\t1323\n",
      "    FLNSIR\t871\n",
      "    GGGGIS\t729\n",
      "    HERPAI\t681\n",
      "    LDSIGK\t664\n",
      "    DGETIR\t615\n",
      "    GEGGVV\t588\n",
      "    NLGAFT\t575\n",
      "    NLSAFT\t555\n",
      "Time for index table init: 0h 0m 1s 537ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 96\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 190804\n",
      "Target db start 1 to 190804\n",
      "[=================================================================] 190.80K 2m 26s 408ms\n",
      "\n",
      "1351.483772 k-mers per position\n",
      "580449 DB matches per sequence\n",
      "5194 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "299 sequences passed prefiltering per query sequence\n",
      "300 median result list length\n",
      "3 sequences with 0 size result lists\n",
      "Time for merging to pref_0: 0h 0m 0s 133ms\n",
      "Time for processing: 0h 2m 30s 361ms\n",
      "align ../data/tmp/18076087456686426716/query ../data/tmp/18076087456686426716/target ../data/tmp/18076087456686426716/search_tmp/2581302743047066002/pref_0 ../data/tmp/18076087456686426716/result --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 48 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 190804 type: Aminoacid\n",
      "Target database size: 190804 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 190.80K 4m 12s 42ms\n",
      "Time for merging to result: 0h 0m 0s 143ms\n",
      "57059312 alignments calculated\n",
      "16632131 sequence pairs passed the thresholds (0.291488 of overall calculated)\n",
      "87.168671 hits per query sequence\n",
      "Time for processing: 0h 4m 13s 958ms\n",
      "rmdb ../data/tmp/18076087456686426716/search_tmp/2581302743047066002/pref_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 157ms\n",
      "rmdb ../data/tmp/18076087456686426716/search_tmp/2581302743047066002/aln_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb ../data/tmp/18076087456686426716/search_tmp/2581302743047066002/input_0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb ../data/tmp/18076087456686426716/search_tmp/2581302743047066002/aln_merge -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "../data3/interim/model_seq_search.txt exists and will be overwritten\n",
      "convertalis ../data/tmp/18076087456686426716/query ../data/tmp/18076087456686426716/target ../data/tmp/18076087456686426716/result ../data3/interim/model_seq_search.txt --sub-mat nucl:nucleotide.out,aa:blosum62.out --format-mode 0 --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov --translation-table 1 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --db-output 0 --db-load-mode 0 --search-type 0 --threads 48 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 190.80K 1s 516ms\n",
      "Time for merging to model_seq_search.txt: 0h 0m 2s 914ms\n",
      "Time for processing: 0h 0m 5s 744ms\n",
      "rmdb ../data/tmp/18076087456686426716/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 169ms\n",
      "rmdb ../data/tmp/18076087456686426716/target -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "rmdb ../data/tmp/18076087456686426716/target_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb ../data/tmp/18076087456686426716/query -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 15ms\n",
      "rmdb ../data/tmp/18076087456686426716/query_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "\n",
      "CPU times: user 50.1 ms, sys: 45.4 ms, total: 95.5 ms\n",
      "Wall time: 6min 56s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "os.system(' '.join(['conda run -n beaker', \n",
    "                    'mmseqs easy-search', \n",
    "                    model_seq_f, \n",
    "                    model_seq_f, \n",
    "                    mmseqs_search_out,\n",
    "                    '../data/tmp',\n",
    "                    '--format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qcov,tcov', \n",
    "                    '--threads 48',\n",
    "                    '-s 7.5']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d9dc7071-8087-44d9-aef5-cb139481e4c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 26 s, sys: 3.43 s, total: 29.4 s\n",
      "Wall time: 29.9 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "mmseqs_search_df = read_mmseqs_results(mmseqs_search_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9b58145-faea-4b51-9f1d-7f349a27b4df",
   "metadata": {},
   "source": [
    "**Query:** test gene\n",
    "\n",
    "**Target:** train gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b065021b-6257-4104-b3a5-e6e78dd4c746",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n"
     ]
    }
   ],
   "source": [
    "top_train_list = list()\n",
    "for fold, fold_test_df in model_seq_info.groupby('test_fold'):\n",
    "    print(fold)\n",
    "    top_mmseqs_df = (mmseqs_search_df[mmseqs_search_df['query'].isin(fold_test_df['seq_id']) & \n",
    "                                      ~mmseqs_search_df['target'].isin(fold_test_df['seq_id'])]\n",
    "                     .sort_values('bits', ascending=False)\n",
    "                     .groupby('query')\n",
    "                     .head(1))\n",
    "    top_train_list.append(top_mmseqs_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c7d7e61b-7b40-46f9-aa58-ba2ce46bfe25",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat_top_hit = pd.concat(top_train_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bab80a4a-4357-42b3-906e-8f7e9cf7ec92",
   "metadata": {},
   "outputs": [],
   "source": [
    "leaked_data = cat_top_hit[((cat_top_hit['qcov'] > 0.8) | \n",
    "                           (cat_top_hit['tcov'] > 0.8)) & \n",
    "                          (cat_top_hit['fident'] > 0.3)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6d7befc4-5a45-42e8-b219-e0f809e860f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_leaked_data = (leaked_data.merge(model_seq_info\n",
    "                                         .rename(columns={'seq_id': 'query'}), \n",
    "                                         how='inner', on='query')\n",
    "                      .merge(model_seq_info\n",
    "                             .rename(columns={'seq_id': 'target'}), \n",
    "                             how='inner', on='target', suffixes=('_query', '_target')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5fd28e60-1e23-483b-a6bc-4fde157538f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>target</th>\n",
       "      <th>qcov</th>\n",
       "      <th>tcov</th>\n",
       "      <th>fident</th>\n",
       "      <th>defensive_query</th>\n",
       "      <th>defensive_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>58716cd0f8d8332725ba61f467c8b195330aeb9267176c9a589765db</td>\n",
       "      <td>27170489b598945be8f2540c4e3d172e6f206c0411643fcb1b7c0eda</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.422</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8d05d5458982cbd3c21f7a0f4b357bfd954940adb13cd5f8a9c2f984</td>\n",
       "      <td>27170489b598945be8f2540c4e3d172e6f206c0411643fcb1b7c0eda</td>\n",
       "      <td>0.990</td>\n",
       "      <td>0.887</td>\n",
       "      <td>0.314</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>774022f6bbc615deab88403dfdaad8ee2ff1a6fbf8cf2da1223cf0ab</td>\n",
       "      <td>27170489b598945be8f2540c4e3d172e6f206c0411643fcb1b7c0eda</td>\n",
       "      <td>0.981</td>\n",
       "      <td>0.698</td>\n",
       "      <td>0.305</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>70c88d578a5af1e89f83c464c7ebc691f638961eef107fb4ea94da6b</td>\n",
       "      <td>27170489b598945be8f2540c4e3d172e6f206c0411643fcb1b7c0eda</td>\n",
       "      <td>0.805</td>\n",
       "      <td>0.047</td>\n",
       "      <td>0.315</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>49cb910efe0fa88b16eb2150960766bf90c3ed91298288b8ececf37d</td>\n",
       "      <td>90722b86e0e71131620cec19bdacc39f959e6ed61186d61420f8cedf</td>\n",
       "      <td>0.703</td>\n",
       "      <td>0.910</td>\n",
       "      <td>0.718</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6005</th>\n",
       "      <td>da598f7cb0033405abd0ce2fa6c5aed08da87e14e2b93085d364348d</td>\n",
       "      <td>aa1a9f256f542b11ffa90a5d1a81fc37a80e7de6d07acb6d90a33b12</td>\n",
       "      <td>0.913</td>\n",
       "      <td>0.223</td>\n",
       "      <td>0.476</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6006</th>\n",
       "      <td>10e33d97fd5d18dccff06414a4a8e8a06e89b3c83feb1bb526fd0adf</td>\n",
       "      <td>e06493b884afe136b6ec5a4c793d02446ef4b306d48e46a5f60bc398</td>\n",
       "      <td>0.889</td>\n",
       "      <td>0.851</td>\n",
       "      <td>0.375</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6007</th>\n",
       "      <td>7a4f38a9fcf54a1da17c4cf3c680695db9494d1898e08832cc6fec25</td>\n",
       "      <td>9492fdd2586010e0d7977dd4964a8e8cf581c6307afd4c508990a9e5</td>\n",
       "      <td>0.885</td>\n",
       "      <td>0.730</td>\n",
       "      <td>0.456</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6008</th>\n",
       "      <td>14ae9af5f3720af41ff4f7fa7784c9e5ece8e2031c0bb91dffde1478</td>\n",
       "      <td>13007df7b3eff484077bf3403ad3daa83642225bdbb9976e696fd8b0</td>\n",
       "      <td>0.952</td>\n",
       "      <td>0.231</td>\n",
       "      <td>0.348</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6009</th>\n",
       "      <td>a0886600f2558eeb111c751df29305683cb3e66e4d780cca0e3d3021</td>\n",
       "      <td>34383861b9750f1381ced8b33f68939ff76c38b5c29281ed954331b9</td>\n",
       "      <td>0.895</td>\n",
       "      <td>0.085</td>\n",
       "      <td>0.470</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6010 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         query   \n",
       "0     58716cd0f8d8332725ba61f467c8b195330aeb9267176c9a589765db  \\\n",
       "1     8d05d5458982cbd3c21f7a0f4b357bfd954940adb13cd5f8a9c2f984   \n",
       "2     774022f6bbc615deab88403dfdaad8ee2ff1a6fbf8cf2da1223cf0ab   \n",
       "3     70c88d578a5af1e89f83c464c7ebc691f638961eef107fb4ea94da6b   \n",
       "4     49cb910efe0fa88b16eb2150960766bf90c3ed91298288b8ececf37d   \n",
       "...                                                        ...   \n",
       "6005  da598f7cb0033405abd0ce2fa6c5aed08da87e14e2b93085d364348d   \n",
       "6006  10e33d97fd5d18dccff06414a4a8e8a06e89b3c83feb1bb526fd0adf   \n",
       "6007  7a4f38a9fcf54a1da17c4cf3c680695db9494d1898e08832cc6fec25   \n",
       "6008  14ae9af5f3720af41ff4f7fa7784c9e5ece8e2031c0bb91dffde1478   \n",
       "6009  a0886600f2558eeb111c751df29305683cb3e66e4d780cca0e3d3021   \n",
       "\n",
       "                                                        target   qcov   tcov   \n",
       "0     27170489b598945be8f2540c4e3d172e6f206c0411643fcb1b7c0eda  0.999  0.999  \\\n",
       "1     27170489b598945be8f2540c4e3d172e6f206c0411643fcb1b7c0eda  0.990  0.887   \n",
       "2     27170489b598945be8f2540c4e3d172e6f206c0411643fcb1b7c0eda  0.981  0.698   \n",
       "3     27170489b598945be8f2540c4e3d172e6f206c0411643fcb1b7c0eda  0.805  0.047   \n",
       "4     90722b86e0e71131620cec19bdacc39f959e6ed61186d61420f8cedf  0.703  0.910   \n",
       "...                                                        ...    ...    ...   \n",
       "6005  aa1a9f256f542b11ffa90a5d1a81fc37a80e7de6d07acb6d90a33b12  0.913  0.223   \n",
       "6006  e06493b884afe136b6ec5a4c793d02446ef4b306d48e46a5f60bc398  0.889  0.851   \n",
       "6007  9492fdd2586010e0d7977dd4964a8e8cf581c6307afd4c508990a9e5  0.885  0.730   \n",
       "6008  13007df7b3eff484077bf3403ad3daa83642225bdbb9976e696fd8b0  0.952  0.231   \n",
       "6009  34383861b9750f1381ced8b33f68939ff76c38b5c29281ed954331b9  0.895  0.085   \n",
       "\n",
       "      fident  defensive_query  defensive_target  \n",
       "0      0.422            False             False  \n",
       "1      0.314            False             False  \n",
       "2      0.305            False             False  \n",
       "3      0.315            False             False  \n",
       "4      0.718            False             False  \n",
       "...      ...              ...               ...  \n",
       "6005   0.476            False             False  \n",
       "6006   0.375            False             False  \n",
       "6007   0.456            False             False  \n",
       "6008   0.348            False             False  \n",
       "6009   0.470            False             False  \n",
       "\n",
       "[6010 rows x 7 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_colwidth', 100)\n",
    "merged_leaked_data[['query', 'target', 'qcov', 'tcov', 'fident', 'defensive_query', 'defensive_target']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8274db62-5af1-49d6-a1e4-040f58c101b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_cat_top_hit = (cat_top_hit[['query', 'target', 'bits']]\n",
    "                      .merge(model_seq_info[['seq_id', 'defensive']]\n",
    "                             .rename(columns={'seq_id': 'target', \n",
    "                                              'defensive': 'target_defensive'}), \n",
    "                             how='inner', on='target'))\n",
    "merged_cat_top_hit['direction'] = merged_cat_top_hit['target_defensive'] * 2 - 1\n",
    "merged_cat_top_hit['prediction'] = merged_cat_top_hit['bits'] * merged_cat_top_hit['direction']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "499c609a-c6e1-4cb1-827c-cb068b674951",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_df = (model_seq_info[['seq_id']]\n",
    "          .merge(merged_cat_top_hit[['query', 'prediction']]\n",
    "                 .rename(columns={'query': 'seq_id'}), \n",
    "                 how='left', on='seq_id'))\n",
    "out_df['prediction'] = out_df['prediction'].fillna(0)\n",
    "out_df['method'] = 'MMseqs nearest neighbor'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "57323f8a-941f-428a-b452-38364484c519",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_df.to_parquet('../data3/interim/cv_predictions_mmseqs.pq', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3576a29-deda-4e36-8cb7-f3d4f4924345",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "beaker",
   "language": "python",
   "name": "beaker"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
